tests/watcher_test.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	client "k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+const (
+	TEST_GLOBAL_SVC_NAME = "echo-test-svc-global"
+	TEST_NS              = "default"
+	TEST_RETRIES         = 5
+)
+
+func GetEndpintSlices() {
+	var kubeconfig *string
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
+	}
+
+	//use the current context in kubeconfig
+	config, err := client.BuildConfigFromFlags("", *kubeconfig)
+
+	fmt.Errorf(err.Error())
+	//Make sure config is not nil
+
+	client, err := kubernetes.NewForConfig(config)
+
+	//Make sure global-svc is created.
+	_, err = client.CoreV1().Services(TEST_NS).Get(context.Background(), TEST_GLOBAL_SVC_NAME, metav1.GetOptions{})
+
+}
+
+func TestGlobalService(t *testing.T) {
```

Comment:
I'm not a big fan of this test. It's unclear how errors are handled, it's unclear if execution can be stopped due to transient errors, the retry loop is pretty erroneous, and there's a lot of set-up involved, precisely why I recommended to use something that is actually used in testing infrastructures. It seems that we use the retry loop to guarantee that we hit both clusters; this will test DNS resolution, not even load balancing.

Here's how I would write this instead:
* Resolve all hostnames and do some regex to extract the cluster name;
* Make an http request and assert connectivity.

We should also follow the common convention of "given, when, then", makes it more readable. If you want to protect against flakiness in CI, then we can write a time based retry loop for the http requests, maybe have it timer based:

```go

func TestFoo(t *testing.T) {
  url := "http://test-foo"
   // resolve DNS, count how many instances we have
  // extract cluster names from instances
  // asserts
  code, err := httpGetWithRetry(...)
   if err != nil {
     t.Fatalf("Unexpected error")
   }
  // assert code is 200
}

func httpGetWithRetry(timeout time.Duration, url) (int, error) {
  resp, err := http.Get(url)
  defer resp.Body.Close()
  if err == nil {
      return resp.StatusCode, nil
  }

  after := time.NewTimer(timeout)
  defer after.Stop()
  retryAfter := time.NewTicker(time.Second)
  defer retryAfter.Stop()
 
  for {
     select {
     case <- after:
         return 500, err

     case <- retryAfter:
          resp, err := http.Get(url)
          defer resp.Body.Close()
          if err == nil {
            return resp.StatusCode, nil
          }
     }
  }
}

Reply:
Agree this looks more neat, thanks for writing this. I will use this and refactor existing code.